src/aead.rs

```diff
@@ -664,6 +664,76 @@ enum Direction {
     Sealing,
 }
 
+trait BitXor {
+    fn bitxor_assign(&mut self, input: [u8; BLOCK_LEN]);
+}
+
+impl BitXor for [u8; BLOCK_LEN] {
+    fn bitxor_assign(&mut self, input: [u8; BLOCK_LEN]) {
+        for (r, a) in self.iter_mut().zip(input.iter()) {
+            *r ^= *a;
+        }
+    }
+}
+
+trait ConvertEndian {
+    fn u64s_be_to_native(&self) -> [u64; 2];
+    fn from_be_u64s(u64s: [u64; 2]) -> Self;
+}
+
+impl ConvertEndian for [u8; BLOCK_LEN] {
+    fn u64s_be_to_native(&self) -> [u64; 2] {
+        [
+            // unwrap safe to use, only errors on length mismatch
+            // lengths guaranteed to match, see [u8]::TryFrom docs
+            u64::from_be_bytes(self[..8].try_into().unwrap()),
+            u64::from_be_bytes(self[8..].try_into().unwrap()),
+        ]
+    }
+
+    fn from_be_u64s(u64s: [u64; 2]) -> Self {
+        let mut out = u64s[0].to_be_bytes().to_vec();
+        out.extend_from_slice(&u64s[1].to_be_bytes());
+
+        // unwrap safe to use, only errors on length mismatch
+        // lengths guaranteed to match, see [u8]::TryFrom docs
+        out.as_slice().try_into().unwrap()
+    }
+}
+
+trait ZeroFrom {
+    fn zero_from(&mut self, index: usize);
+}
+
+impl ZeroFrom for [u8; BLOCK_LEN] {
+    fn zero_from(&mut self, index: usize) {
+        // panics when index > self.len()
+        // TODO: change the API to return Result<(), Error>, and bubble up?
+        let (_, zero) = self.split_at_mut(index);
+        for elem in zero.iter_mut() {
+            *elem = 0u8;
+        }
+    }
+}
+
+trait Overwrite {
+    /// Overwrite part of a slice
+    ///
+    /// Panics:
+    ///
+    /// Panics if index > self.len()
+    // TODO: change API to return Result<(), Error>, and bubble up?
+    fn overwrite_part_at(&mut self, index: usize, a: &[u8]);
```

Comment:
I understand that you factored each part of `Block` into a separate trait and then implemented them all for `[u8; BLOCK_LEN]` but I think that made things more complicated than necessary. When we get rid of uses of `Block` I expect the code should actually get simpler, not bigger.

For example, I suspect that the right thing to do is just completely rewrite the code that uses `overwrite_part_at` to be simpler/smarter.

Reply:
> For example, I suspect that the right thing to do is just completely rewrite the code that uses overwrite_part_at to be simpler/smarter.

Think you're right. Going from the bottom up (replacing in `aead/aes`, breaking things, fixing things), overall design got lost. Think I will have a better feel for design going from top uses down. Will keep the simplification factor in mind.